File: backend/internal/models/session.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type PracticeSession struct {
	ID                   uuid.UUID              `json:"id" db:"id"`
	UserID               uuid.UUID              `json:"user_id" db:"user_id"`
	ProgramID            uuid.UUID              `json:"program_id" db:"program_id"`
	ProgramName          *string                `json:"program_name,omitempty"`
	StartedAt            time.Time              `json:"started_at" db:"started_at"`
	CompletedAt          *time.Time             `json:"completed_at,omitempty" db:"completed_at"`
	TotalDurationSeconds *int                   `json:"total_duration_seconds,omitempty" db:"total_duration_seconds"`
	CompletionRate       *float64               `json:"completion_rate,omitempty" db:"completion_rate"`
	Notes                *string                `json:"notes,omitempty" db:"notes"`
	DeviceInfo           map[string]interface{} `json:"device_info,omitempty" db:"device_info"`
}

type ExerciseLog struct {
	ID                     uuid.UUID  `json:"id" db:"id"`
	SessionID              uuid.UUID  `json:"session_id" db:"session_id"`
	ExerciseID             *uuid.UUID `json:"exercise_id,omitempty" db:"exercise_id"`
	StartedAt              *time.Time `json:"started_at,omitempty" db:"started_at"`
	CompletedAt            *time.Time `json:"completed_at,omitempty" db:"completed_at"`
	PlannedDurationSeconds *int       `json:"planned_duration_seconds,omitempty" db:"planned_duration_seconds"`
	ActualDurationSeconds  *int       `json:"actual_duration_seconds,omitempty" db:"actual_duration_seconds"`
	RepetitionsPlanned     *int       `json:"repetitions_planned,omitempty" db:"repetitions_planned"`
	RepetitionsCompleted   *int       `json:"repetitions_completed,omitempty" db:"repetitions_completed"`
	Skipped                bool       `json:"skipped" db:"skipped"`
	Notes                  *string    `json:"notes,omitempty" db:"notes"`
}

type SessionWithLogs struct {
	Session      PracticeSession `json:"session"`
	ExerciseLogs []ExerciseLog   `json:"exercise_logs"`
}

type SessionStats struct {
	TotalSessions         int     `json:"total_sessions"`
	CompletedSessions     int     `json:"completed_sessions"`
	TotalDurationMinutes  int     `json:"total_duration_minutes"`
	AverageCompletionRate float64 `json:"average_completion_rate"`
	CurrentStreak         int     `json:"current_streak"`
	LongestStreak         int     `json:"longest_streak"`
}

File: app/lib/models/session.dart
class PracticeSession {
  final String id;
  final String userId;
  final String programId;
  final String? programName;
  final DateTime startedAt;
  final DateTime? completedAt;
  final int? totalDurationSeconds;
  final double? completionRate;
  final String? notes;

  PracticeSession({
    required this.id,
    required this.userId,
    required this.programId,
    this.programName,
    required this.startedAt,
    this.completedAt,
    this.totalDurationSeconds,
    this.completionRate,
    this.notes,
  });

  factory PracticeSession.fromJson(Map<String, dynamic> json) {
    return PracticeSession(
      id: json['id'],
      userId: json['user_id'],
      programId: json['program_id'],
      programName: json['program_name'],
      startedAt: DateTime.parse(json['started_at']),
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'])
          : null,
      totalDurationSeconds: json['total_duration_seconds'],
      completionRate: json['completion_rate']?.toDouble(),
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'program_id': programId,
      'program_name': programName,
      'started_at': startedAt.toIso8601String(),
      'completed_at': completedAt?.toIso8601String(),
      'total_duration_seconds': totalDurationSeconds,
      'completion_rate': completionRate,
      'notes': notes,
    };
  }
}

class ExerciseLog {
  final String? id;
  final String sessionId;
  final String? exerciseId;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final int? plannedDurationSeconds;
  final int? actualDurationSeconds;
  final int? repetitionsPlanned;
  final int? repetitionsCompleted;
  final bool skipped;
  final String? notes;

  ExerciseLog({
    this.id,
    required this.sessionId,
    this.exerciseId,
    this.startedAt,
    this.completedAt,
    this.plannedDurationSeconds,
    this.actualDurationSeconds,
    this.repetitionsPlanned,
    this.repetitionsCompleted,
    this.skipped = false,
    this.notes,
  });

  factory ExerciseLog.fromJson(Map<String, dynamic> json) {
    return ExerciseLog(
      id: json['id'],
      sessionId: json['session_id'],
      exerciseId: json['exercise_id'],
      startedAt: json['started_at'] != null ? DateTime.parse(json['started_at']) : null,
      completedAt: json['completed_at'] != null ? DateTime.parse(json['completed_at']) : null,
      plannedDurationSeconds: json['planned_duration_seconds'],
      actualDurationSeconds: json['actual_duration_seconds'],
      repetitionsPlanned: json['repetitions_planned'],
      repetitionsCompleted: json['repetitions_completed'],
      skipped: json['skipped'] ?? false,
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'session_id': sessionId,
      'exercise_id': exerciseId,
      'started_at': startedAt?.toIso8601String(),
      'completed_at': completedAt?.toIso8601String(),
      'planned_duration_seconds': plannedDurationSeconds,
      'actual_duration_seconds': actualDurationSeconds,
      'repetitions_planned': repetitionsPlanned,
      'repetitions_completed': repetitionsCompleted,
      'skipped': skipped,
      'notes': notes,
    };
  }
}

class SessionWithLogs {
  final PracticeSession session;
  final List<ExerciseLog> exerciseLogs;

  SessionWithLogs({
    required this.session,
    required this.exerciseLogs,
  });

  factory SessionWithLogs.fromJson(Map<String, dynamic> json) {
    return SessionWithLogs(
      session: PracticeSession.fromJson(json['session']),
      exerciseLogs: (json['exercise_logs'] as List<dynamic>?)
              ?.map((log) => ExerciseLog.fromJson(log))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'session': session.toJson(),
      'exercise_logs': exerciseLogs.map((log) => log.toJson()).toList(),
    };
  }
}
